system/lib/compiler-rt/lib/lsan/lsan_common_emscripten.cc

```diff
@@ -77,15 +77,14 @@ void InitializePlatformSpecificModules() {
   }
 }
 
-extern "C" {
-  extern int __data_end;
-  extern int __heap_base;
-}
-
 // Scans global variables for heap pointers.
 void ProcessGlobalRegions(Frontier *frontier) {
   if (!flags()->use_globals) return;
-  ScanGlobalRange(0, (uptr) &__data_end, frontier);
+  ScanGlobalRange(EM_ASM_INT({
+    return GLOBAL_BASE;
+  }), EM_ASM_INT({
+    return STACK_TOP;
+  }), frontier);
```

Comment:
How is this related to PIC? Does the behavior change in the non-PIC case? It's unfortunate to have a tight dependency on Emscripten in code that we would ideally upstream at some point.

Reply:
This is the emscripten-specific implementation for this function. The behaviour does not change in the non-PIC case here, as we are simply using alternatives to `__data_end`.